server/tables/pgcatalog/inmem_index.go

```diff
@@ -241,3 +232,108 @@ func (p *inMemIndexPartIter) Next(context *sql.Context) (sql.Partition, error) {
+// unique and non-unique indexes.
+type inMemIndexStorage[T any] struct {
+	uniqTree    *btree.BTreeG[T]
+	nonUniqTree *btree.BTreeG[[]T]
+}
+
+// NewUniqueInMemIndexStorage creates a new in-memory index storage for a unique index.
+func NewUniqueInMemIndexStorage[T any](lessFunc func(a, b T) bool) *inMemIndexStorage[T] {
+	return &inMemIndexStorage[T]{
+		uniqTree: btree.NewG[T](2, lessFunc),
+	}
+}
+
+// NewNonUniqueInMemIndexStorage creates a new in-memory index storage for a non-unique index.
+func NewNonUniqueInMemIndexStorage[T any](lessFunc func(a, b []T) bool) *inMemIndexStorage[T] {
+	return &inMemIndexStorage[T]{
+		nonUniqTree: btree.NewG[[]T](2, lessFunc),
+	}
+}
+
+// Add adds a value to the in-memory index storage.
+func (s *inMemIndexStorage[T]) Add(val T) {
+	if s.uniqTree != nil {
+		s.uniqTree.ReplaceOrInsert(val)
+	} else {
+		existing, replaced := s.nonUniqTree.ReplaceOrInsert([]T{val})
+		if replaced {
+			existing = append(existing, val)
+			s.nonUniqTree.ReplaceOrInsert(existing)
+		}
+	}
+}
+
+// IterRange implements an in-order iteration over the index values in the given range, inclusive. All values in the
+// index in the range are sent to the channel
+func (s *inMemIndexStorage[T]) IterRange(gte, lte T, c chan T) {
+	if s.uniqTree != nil {
+		s.uniqTree.AscendRange(gte, lte, s.sendItem(c))
+	} else {
+		s.nonUniqTree.AscendRange([]T{gte}, []T{lte}, s.sendItems(c))
+	}
+
+	s.iterKey(lte, c)
+}
+
+// IterGreaterThanEqual implements an in-order iteration over the index values greater than or equal to the given value.
+// All values in the index greater than or equal to the given value are sent to the channel.
+func (s *inMemIndexStorage[T]) IterGreaterThanEqual(gte T, c chan T) {
+	if s.uniqTree != nil {
+		s.uniqTree.AscendGreaterOrEqual(gte, s.sendItem(c))
+	} else {
+		s.nonUniqTree.AscendGreaterOrEqual([]T{gte}, s.sendItems(c))
+	}
+}
+
+// IterLessThan implements an in-order iteration over the index values less than or equal to the given value.
+// All values in the index less than or equal to the given value are sent to the channel.
+func (s *inMemIndexStorage[T]) IterLessThan(lte T, c chan T) {
+	if s.uniqTree != nil {
+		s.uniqTree.AscendLessThan(lte, s.sendItem(c))
+	} else {
+		s.nonUniqTree.AscendLessThan([]T{lte}, s.sendItems(c))
+	}
+
+	s.iterKey(lte, c)
+}
+
+// iterKey sends the value for the given key to the channel if it exists in the index.
+// This is used to include the upper bound of a range scan, since the btree package uses a half-open range in all of
+// its Ascend methods.
+func (s *inMemIndexStorage[T]) iterKey(v T, c chan T) {
+	if s.uniqTree != nil {
+		val, ok := s.uniqTree.Get(v)
+		if ok {
+			c <- val
+		}
+	} else {
+		vals, ok := s.nonUniqTree.Get([]T{v})
+		if ok {
+			for _, val := range vals {
+				c <- val
+			}
+		}
+	}
+}
+
+// sendItem returns a function that sends the given item to the channel.
+func (s *inMemIndexStorage[T]) sendItem(c chan T) func(item T) bool {
```

Comment:
The function returns a `bool`, but it's not clear why it returns a `bool`. I can see that it's due to the index interface, but I think we could expand the comment here to make it a bit more straightforward (same for `sendItems` too).